imu_filter_madgwick/src/imu_filter_ros.cpp

```diff
@@ -517,3 +539,29 @@ void ImuFilterRos::checkTimeJump()
 
     reset();
 }
+
+bool ImuFilterRos::checkLargeTimeJumpForward(const ros::Time& imu_stamp)
+{
+    bool largeJump = false;
+    if (constant_dt_ > 0.0)
+    {
+        return largeJump;  // use constant dt
+    }
+
+    if (imu_stamp > last_time_ + forward_large_time_jump_threshold_ &&
+        !last_time_.isZero())
+    {
+        ROS_ERROR(
+            "Detected large jump forward, jump is %f[s]. rejecting measurement",
+            (imu_stamp - last_time_).toSec());
+        largeJump = true;
+        last_time_ = imu_stamp;  // save time for next move
+        if (reset_large_forward_time_jump_)
```

Comment:
I wonder if we shouldn't always reset. This would simplify the code a lot:

* we can remove the `reset_large_forward_time_jump` parameter
* we can probably even remove the return value from this function, and the `largeJump` variable
* we can make this function very similar to `checkTimeJump()`, and perhaps even merge this function into it.

Also, it sounds safer to always reset. The current implementation only rejects the first measurement after a large jump forward, but if the filter isn't reset, the next message will still have a large jump in values with a small dt, so rejecting the first measurement doesn't help at all.

Reply:
I don't think we should always reset. In our application, we experience one large time jump when our robot establishes communication. It could be very dangerous to reset the IMU suddenly. Instead, I believe it is better to ignore one measurement. Why should the next message have a large jump in values?

Reply:
Ah yes, after reading #160 again I remember the dilemma in your application. There can be two sources of large forward time jumps between two consecutive messages:

1. **Pause**: If the connection to the IMU was lost somehow and messages were dropped for an extensive time, when the first message arrives again you get a large time jump *and* a large jump in values (because the IMU was moved in the meantime).
2. **Clock update**: There is no connection error, and the IMU messages arrive continuously at the same rate, but the clock was moved forward between two messages. There's a large time jump but *no* large jump in values.

You're trying to deal with case 2 here, but I was thinking about case 1.

As I already wrote in https://github.com/CCNYRoboticsLab/imu_tools/issues/160#issuecomment-1127545462, I think it's a bad idea to try to handle case 2 in the imu_filter_madgwick.

The source of the problem is the incorrect system clock. You are feeding bad data with incorrect time stamps (and a time jump) to all downstream software (such as imu_filter_madgwick). This will cause bad behavior in hundreds of different places, often with very subtle effects that will be a nightmare to track down.

Instead of trying to make the downstream software deal with incorrect inputs, you should fix the problem at the source and provide correct time stamps. Two wrongs don't make a right.

There are many ways you could fix the time jumps:

1. Add a real time clock to the robot (they only cost a couple of bucks).
2. Don't update the time after connecting to the internet.
3. ...